test-data/unit/check-generics.test

```diff
@@ -2477,3 +2478,28 @@ b: B
 
 reveal_type([a, b])  # N: Revealed type is "builtins.list[typing.Sequence*[builtins.object]]"
 [builtins fixtures/list.pyi]
+
+[case testGenericJoinRecursiveInvariant]
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+class I(Generic[T]): ...
+
+class A(I[A]): ...
+class B(I[B]): ...
+
+a: A
+b: B
+reveal_type([a, b])  # N: Revealed type is "builtins.list[builtins.object*]"
+[builtins fixtures/list.pyi]
+
+[case testGenericJoinNestedInvariantAny]
+from typing import Any, Generic, TypeVar
+
+T = TypeVar("T")
+class I(Generic[T]): ...
+
+a: I[int]
+b: I[Any]
+reveal_type([a, b])  # N: Revealed type is "builtins.list[__main__.I*[Any]]"
```

Comment:
Can you also add `reveal_type([b, a])` to (hopefully) test the reversed operand order.